crates/consensus/src/crypto.rs

```diff
@@ -286,4 +388,92 @@ mod tests {
 
         assert_eq!(secp256k1_recovered, k256_recovered);
     }
+
+    #[cfg(feature = "crypto-backend")]
+    mod backend_tests {
+        use super::*;
+        use crate::crypto::backend::{CryptoProvider, install_default_provider, try_get_provider};
+        use alloy_primitives::{Address, B256, Signature};
+
+        /// Mock crypto provider for testing
+        struct MockCryptoProvider {
+            should_fail: bool,
+        }
+
+        impl CryptoProvider for MockCryptoProvider {
+            type Error = RecoveryError;
+
+            fn recover_signer_unchecked(
+                &self,
+                _sig: &[u8; 65],
+                _msg: &[u8; 32],
+            ) -> Result<Address, Self::Error> {
+                if self.should_fail {
+                    Err(RecoveryError::new())
+                } else {
+                    // Return a known/fixed test address
+                    // We don't care about the implementation specific details
+                    // of a particular CryptoProvider
+                    Ok(Address::from([0x42; 20]))
+                }
+            }
+        }
+
+        #[test]
+        fn test_provider_already_set_error() {
+            // First installation might work or fail if already set from another test
+            // Since tests are ran in parallel.
+            let provider1 = Arc::new(MockCryptoProvider { should_fail: false });
+            let result1 = crate::crypto::backend::install_default_provider(provider1);
+            
+            // Second installation should always fail since OnceLock can only be set once
+            let provider2 = Arc::new(MockCryptoProvider { should_fail: true });
+            let result2 = crate::crypto::backend::install_default_provider(provider2.clone());
+            
+            // The second attempt should fail with ProviderAlreadySetError
+            assert!(result2.is_err());
+            
+            // The error should contain the provider we tried to install (provider2)
+            if let Err(err) = result2 {
+                // Verify the returned provider is the same as the one we tried to install
+                assert!(Arc::ptr_eq(&err.provider, &provider2));
+            }
+        }
+
+
+        /// Test that when crypto-backend feature is enabled but no provider is installed,
+        /// it falls back to the default implementation
+        /// 
+        /// Note: this test is unreliable because if the CryptoProvider is set, it will use the MockProvider
+        /// which doesn't have ideal behaviour
```

Comment:
Noting this note. We should probably remove this since we cannot guarantee that another test will not set the CryptoProvider

Reply:
removed in d106417b1e441c97faf1ff675c6ad258387c645a